react/AppIcon/index.jsx

```diff
@@ -7,16 +7,129 @@ import Icon from '../Icon'
 
 import appDefaultIcon from '../../assets/icons/ui/cube.svg'
 
+let domain
+const _getDomain = () => {
+  if (domain) return domain
+  const root = document.querySelector('[role=application]')
+  if (!root) {
+    throw new Error('Cannot find [role=application] element')
+  }
+  domain = root.dataset.cozyDomain
+  if (!domain) {
+    throw new Error(
+      'Application element does not contain `data-cozy-domain` attribute'
+    )
+  }
+  return domain
+}
+
+let protocol
+const _getProtocol = () => {
+  if (protocol) return protocol
+  protocol = window.location.protocol
+  return protocol
+}
+
+const _preload = url => {
+  return new Promise((resolve, reject) => {
+    const loader = document.createElement('img')
+    loader.onload = () => {
+      resolve(url)
+    }
+    loader.onerror = () => {
+      reject(new Error(`Error while preloading ${url}`))
+    }
+    loader.src = url
+  })
+}
+
+const _getAppIconURL = app => {
+  let path = _getInstalledIconPath(app)
+  path = path || _getRegistryIconPath(app)
+  return path ? `${_getProtocol()}//${_getDomain()}${path}` : null
+}
+
+const _getInstalledIconPath = app => {
+  return app.links && app.links.icon
+}
+
+const _getRegistryIconPath = app => {
+  return (
+    app.latest_version &&
+    app.latest_version.version &&
+    `/registry/${app.slug}/${app.latest_version.version}/icon`
+  )
+}
+
+const _loaded = []
```

Comment:
I think `_loaded` should be managed by the `_preload` function. This way, AppIcon does not have to do it. A cache abstraction could maybe help to divide responsibilities more cleanly.

Reply:
The only flaw I see is that the AppIcon, in the constructor, needs to know if a icon has already been loaded or not, to renders it immediately without a temporary `fetching` state.

Reply:
Actually, we still can use in the constructor.